drivers/thermal/thermal_cpufreq_cooling.c

```diff
@@ -0,0 +1,234 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/cpufreq.h>
+#include <nuttx/kmalloc.h>
+
+#include "thermal_core.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct cpufreq_cooling_device_s
+{
+  FAR const struct cpufreq_frequency_table *table;
+  FAR struct cpufreq_policy *policy;
+  FAR struct cpufreq_qos *qos;
+  unsigned int cur_state;
+  unsigned int max_state;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int cpufreq_get_max_state(FAR struct thermal_cooling_device_s *cdev,
+                                 FAR unsigned int *state);
+static int cpufreq_get_state    (FAR struct thermal_cooling_device_s *cdev,
+                                 FAR unsigned int *state);
+static int cpufreq_set_state    (FAR struct thermal_cooling_device_s *cdev,
+                                 unsigned int state);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct thermal_cooling_device_ops_s g_cpufreq_cdev_ops =
+{
+  .set_state     = cpufreq_set_state,
+  .get_state     = cpufreq_get_state,
+  .get_max_state = cpufreq_get_max_state,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int cpufreq_get_max_state(FAR struct thermal_cooling_device_s *cdev,
+                                 FAR unsigned int *state)
+{
+  struct cpufreq_cooling_device_s *cpufreq_cdev = cdev->devdata;
+
+  *state = cpufreq_cdev->max_state;
+  return OK;
+}
+
+static int cpufreq_get_state(FAR struct thermal_cooling_device_s *cdev,
+                             FAR unsigned int *state)
+{
+  struct cpufreq_cooling_device_s *cpufreq_cdev = cdev->devdata;
+
+  *state = cpufreq_cdev->cur_state;
+  return OK;
+}
+
+static int cpufreq_set_state(FAR struct thermal_cooling_device_s *cdev,
+                             unsigned int state)
+{
+  struct cpufreq_cooling_device_s *cpufreq_cdev = cdev->devdata;
+  unsigned int index = cpufreq_cdev->max_state - state;
```

Comment:
If `state > max_state` then `index` will be out of bounds of `cpufreq_cdev->table`. Is there some logic elsewhere to prevent this?

(If not, can `index` be clamped to the range 0 through max_state (inclusive) here?)

Reply:
> If `state > max_state` then `index` will be out of bounds of `cpufreq_cdev->table`. Is there some logic elsewhere to prevent this?
> 
> (If not, can `index` be clamped to the range 0 through max_state (inclusive) here?)

It`s updated by core/governor, please simply refer to [drivers/thermal/thermal_step_wise.c +54](https://github.com/apache/nuttx/pull/12824/files/d40b702de07c7e3af5af1a1dbe58c35d6b66498f#diff-b9e4ae8c8db0ca7b37552a812f5d882584c5e99079aad771c0ff7158ef32f219R54).
The currently dependent module - “cpufreq” has not been uploaded yet. This PR will be updated in the near future to resolve the dependencies and add detailed instructions.